File: src/semana-2/Kth-Smallest-Element-in-a-BST.cpp
#include <bits/stdc++.h>
using namespace std;
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/
// 230. Kth Smallest Element in a BST
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        auto aux = root;
        stack<TreeNode*> tree;
        tree.push(root);
        int node;
        
        while(root->left != NULL){
            aux = root->left;
            root = aux;
            tree.push(root);
        }
        
        for(int i = 0; i < k; i++){
            root = tree.top();
            tree.pop();
            node = root->val;
            
            if(root->right != NULL){
                aux = root->right;
                root = aux;
                tree.push(root);
                
                while(root->left != NULL){
                    aux = root->left;
                    root = aux;
                    tree.push(root);
                } // while
            }// if
        }// for
        return node;
    }
};
int main(){
    
    return 0;
}

File: src/semana-2/JavaScript/Kth-Smallest-Element-in-a-BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 * https://leetcode.com/problems/kth-smallest-element-in-a-bst/
 */
const root = [3,1,4,null,2];	
const k = 1;
var kthSmallest = (root, k) => {
  var stack = [];
  var node;
  var nodeVal;
  stack.push(root);
  
  while (root.left) {
    node = root.left;
    root = node;
    stack.push(root);
  }

  while (k--) {
    root = stack.pop();
    nodeVal = root.val;

    if (root.right) {
      node = root.right;
      root = node;
      stack.push(root);

      while (root.left) {
        node = root.left;
        root = node;
        stack.push(root);
      }
    }
  }
  return nodeVal;
};

console.log(kthSmallest(root, k));